### DISCUSSION TITLE
[Proposal]: Allow pattern matching to assign to new variable in more cases

### DISCUSSION THREAD

**Discussion by OJacot-Descombes** [dotnet/csharplang#7945] [Language Ideas]
Given the declaration
```
class C
{
    public string? P { get; set; }
}
```
we can write

```
C c = new();

if (c is { P: { } s }) { }         // OK
```
This tests the property `P` for being not null and assigns its value to the new variable `s` if the result is `true`. This is not very readable. It would be more readable if we could write:
```
if (c is { P: not null s }) { }    // Does not compile
```
There are also other cases where assigning a new variable would be handy:
```
if (c is { P.Length: > 0 i }) { }  // Does not compile
if (c is { P: "x" or "y" s }) { }  // Does not compile
```

Proposal: allow pattern matching to test an expression and conditionally declare a new variable to the result of that expression in more cases.

**Comment by 333fred** [comment]
Note that in all of these cases, you can just append `and var s` to declare a variable (the `or` case could wrap that part in parentheses). I'm generally unconvinced that removing the need for an `and` pattern results in code that is easier to understand.

**Reply by OJacot-Descombes** [reply]
Good point. `not null and var s` is more explicit than `{ } s` but looks very convoluted. I would still prefer the latter in this case.

**Reply by Unknown6656** [reply]
@OJacot-Descombes or, you could simply do:
```csharp
if (c is { P: string s })
{
    ...
}
```


https://sharplab.io/#v2:C4LgTgrgdgPgAgJgIwFgBQ70GEAEBjHAXhygFMB3ACiggBtaBKAbnQHpX8iSLKAiYUgGdgvZjnY4APAFppOaHgD2AW2WkowHADNFYHAE8AhvoCEOEA0wY0ASy05KBG4JwBvHAAUQOOEgAMOC4AvpZoOOE+SACclAAkvEb65gxugkGiLGiktIKk6BGRMfFQigAOpOZ8zFboYKRKYAAmOFiUvn4A/J7MQA

**Reply by OJacot-Descombes** [reply]
Sure, you can write `c is { P: string s }` or `c is { P: object s }`, but this is a type test and it does not clearly express the intent of testing for `not null`.

**Comment by TahirAhmadov** [comment]
> There are also other cases where assigning a new variable would be handy:

These examples look a bit strange to me, it's not as easy to read.

> `if (c is { P: not null s }) { }    // Does not compile`

This would be great though.

**Reply by OJacot-Descombes** [reply]
Does `if (c is { P: { } s }) { }` look less strange and easier to read to you? (We are not talking about the "strangeness" of pattern matching in general. We are comparing different ways of expressing the same pattern matching test.)
